runtime/src/main/java/org/corfudb/runtime/view/stream/StreamAddressSpace.java

```diff
@@ -1,57 +1,89 @@
+import java.io.DataOutput;
+import java.io.DataOutputStream;
+import java.io.IOException;
 import java.util.ArrayList;
+import java.util.Collection;
 import java.util.Collections;
 import java.util.List;
 import java.util.NavigableSet;
 import java.util.TreeSet;
 
 /**
  * This class represents the space of all addresses belonging to a stream.
- *
+ * <p>
  * A stream's address space is defined by:
- *       1. The collection of all addresses that belong to this stream.
- *       2. The trim mark (last trimmed address, i.e., an address that is no longer present and that was subsumed by
- *       a checkpoint).
-  *
+ * 1. The collection of all addresses that belong to this stream.
+ * 2. The trim mark (last trimmed address, i.e., an address that is no longer present and that was subsumed by
+ * a checkpoint).
+ * <p>
  * Created by annym on 03/06/2019
  */
 @Slf4j
-public class StreamAddressSpace {
-
-    private static final int NO_ADDRESSES = 0;
+final public class StreamAddressSpace {
 
     // Holds the last trimmed address for this stream.
     // Note: keeping the last trimmed address is required in order to properly set the stream tail on sequencer resets
     // when a stream has been checkpointed and trimmed and there are no further updates to this stream.
     private long trimMark;
 
     // Holds the complete map of addresses for this stream.
-    private final Roaring64NavigableMap addressMap;
+    // TODO(Maithem): This container should be of a higher abstraction
+    private final Roaring64NavigableMap bitmap;
 
-    public StreamAddressSpace(long trimMark, Roaring64NavigableMap addressMap) {
+    /**
+     * This constructor is required to facilitate deserialization, keep it provide.
+     * The internal bitmap container shouldn't be exposed to external consumers.
+     * @param trimMark Stream's trim mark
+     * @param bitmap Stream's bitmap
+     */
+    public StreamAddressSpace(long trimMark, Roaring64NavigableMap bitmap) {
         this.trimMark = trimMark;
-        this.addressMap = addressMap;
+        this.bitmap = bitmap;
+        //TODO(Maithem): re-enable this. cluster re-configuration tests are failing because
+        // they relay on incorrect behavior.
+        //this.trim(trimMark);
     }
 
     public StreamAddressSpace() {
-        this.addressMap = Roaring64NavigableMap.bitmapOf();
-        this.trimMark = Address.NON_ADDRESS;
+        this(Address.NON_ADDRESS, Roaring64NavigableMap.bitmapOf());
     }
 
-    public Roaring64NavigableMap getAddressMap() {
-        return addressMap;
+    public StreamAddressSpace(long trimMark, Collection<Long> addresses) {
+        this(trimMark, Roaring64NavigableMap.bitmapOf());
+        addresses.forEach(bitmap::addLong);
+        //this.trim(trimMark);
+    }
+
+    public StreamAddressSpace(Collection<Long> addresses) {
+        this(Address.NON_ADDRESS, addresses);
+    }
+
+    public boolean contains(long address) {
+        // TODO(Maithem): < or <= ?
```

Comment:
If the bitmap = [1 , 2, 3] and setTrimMark = 2 then contains would return true for 1 and 2, but that is incorrect. setTrimMark should truncate the bitmap, but there is existing code that depends on this incorrect behavior.

Reply:
since the bitmap implementation accepts negative numbers and interprets them as greater than positive numbers, we need to return false for negative numbers. For example, the bitmap implementation has this order 0, 1 , ..., Long.Max_value, Long.min_value, ... -2, -1